New checks for `admin check` command

kevinrr888 opened the pull request:
New checks for `admin check` command

- Implemented SYSTEM_CONFIG check:
	- Checks ZooKeeper locks for Accumulo server processes
	- Checks ZooKeeper table nodes
	- Checks that the WAL metadata in ZooKeeper is valid
- Added new SERVER_CONFIG check:
	- Checks that all configured properties are valid
	- Checks that required properties are present in the config
- Added new tests in `AdminCheckIT` for SYSTEM_CONFIG and SERVER_CONFIG. Added failing test cases for all checks.
- Updated description for #4892. This describes all the checks thus far for the new `accumulo admin check` command. See this for the complete functionality of the command.
- Considered adding check to MetadataConstraints to ensure tablets reference files that actually exist (i.e., in `MetadataConstraints.validateDataFileMetadata`, check that `stf.getPath()` exists in HDFS). This would make sure the file exists before the mutation is ever written and would also check that it exists when we run the admin check command for metadata (root metadata, root table, and metadata table): see `MetadataCheckRunner.checkColumns()`. However, this led to a ComprehensiveIT test failure, so I assume there's a reason we're not checking its existance in MetadataConstraints. Would this be something we want? Or is there a reason we are not checking for this?
- Deleted CheckServerConfig (run via `accumulo check-server-config`) as the new `accumulo admin check run SERVER_CONFIG` will inherently do the same checks. Maybe we don't want to completely get rid of this check yet. Maybe it would be better to have the check just output that the check has moved under `accumulo admin check`?
- There were a couple other existing check commands that I originally considered moving under this new admin command (CheckAccumuloProperties and CheckCompactionConfig), however, their usage is specifically to check a file irrespective of any running Accumulo instance, so I did not feel it would make sense to move these under the new admin command.

This PR along with #4957 closes #4892

kevinrr888 commented on server/base/src/main/java/org/apache/accumulo/server/util/checkCommand/SystemConfigCheckRunner.java:
```diff
@@ -18,22 +18,230 @@
+          }
+          for (String compactor : compactors) {
+            // for each running compactor, ensure a zk lock exists for it
+            boolean checkedLock = false;
+            String compactorQueuesPath = zkRoot + Constants.ZCOMPACTORS;
+            var compactorQueues = zrw.getChildren(compactorQueuesPath);
+            // find the queue the compactor is in
+            for (var queue : compactorQueues) {
+              String compactorQueuePath = compactorQueuesPath + "/" + queue;
+              String lockPath = compactorQueuePath + "/" + compactor;
+              if (zrw.exists(lockPath)) {
+                status = checkLock(lockPath, proc, true, zs, status);
+                checkedLock = true;
+                break;
+              }
+            }
+            if (!checkedLock) {
+              log.warn("Did not find a ZooKeeper lock for the compactor {}!", compactor);
+              status = Admin.CheckCommand.CheckStatus.FAILED;
+            }
+          }
+          break;
+        case MONITOR:
+          // nonessential process
+          status = checkLock(zkRoot + Constants.ZMONITOR_LOCK, proc, false, zs, status);
+          break;
+        case SCAN_SERVER:
+          // nonessential process(es)
+          if (sservers.isEmpty()) {
+            log.debug("No scan servers appear to be running... This may or may not be expected");
+          }
+          for (String sserver : sservers) {
+            status =
+                checkLock(zkRoot + Constants.ZSSERVERS + "/" + sserver, proc, true, zs, status);
+          }
+          break;
+        default:
+          throw new IllegalStateException("Unhandled case: " + proc);
+      }
+    }
```
I imagine some of the non essential processes here in 3.x are essential in 4.x. I believe COMPACTION_COORDINATOR, COMPACTOR, SCAN_SERVER are all essential in 4.x? Is this correct? Will need to change this when merging into 4.x if so.

kevinrr888 commented on server/base/src/main/java/org/apache/accumulo/server/util/checkCommand/SystemConfigCheckRunner.java:
```diff
@@ -18,22 +18,230 @@
+    final String zkRoot = context.getZooKeeperRoot();
+    final var zs = context.getZooSession();
+    final var zrw = zs.asReaderWriter();
+    final var compactors = context.instanceOperations().getCompactors();
+    final var sservers = context.instanceOperations().getScanServers();
```
Looking back at this code again: this gets the compactors and sservers from scanning ZK then validating the lock node... I hadn't considered how this was gathering the data. Thinking most or all of the code for the cases for COMPACTOR and SCAN_SERVER below can be removed. Thinking the new cases would look something like:
```
case COMPACTOR:
	var compactors = context.instanceOperations().getCompactors();
	if (compactors.isEmpty()) {
		// warn user (debug) that no compactors were found to be running which may or may not be expected.
	}
	break;
```

kevinrr888 commented on server/base/src/main/java/org/apache/accumulo/server/util/checkCommand/SystemConfigCheckRunner.java:
```diff
@@ -18,22 +18,230 @@
+        }
+      }
+    }
+    if (!tserverInstances.equals(seenTServerInstancesAtWals)) {
+      log.warn(
+          "Expected WAL metadata in ZooKeeper for all tservers. tservers={} tservers seen storing WAL metadata={}",
+          tserverInstances, seenTServerInstancesAtWals);
+      status = Admin.CheckCommand.CheckStatus.FAILED;
+    }
```
Should we expect this? If no writes have occurred yet for the TServer, this check would fail, but I imagine in any normal scenario this check is fine.

kevinrr888 replied:
Okay, what should be done about this check? Could leave as is, retry for a period of time before failing, something else?

kevinrr888 replied:
Removed in 1af0ad37e051d4f6c770110e711fc0d122cb832f

keith-turner commented on server/base/src/main/java/org/apache/accumulo/server/util/checkCommand/ServerConfigCheckRunner.java:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    log.trace("Checking that all required config properties are present");
+    // there are many properties that should be set (default value or user set), identifying them
+    // all and checking them here is unrealistic. Some property that is not set but is expected
+    // will likely result in some sort of failure eventually anyway. We will just check a few
+    // obvious required properties here.
+    Set<Property> requiredProps = Set.of(Property.INSTANCE_ZK_HOST, Property.INSTANCE_ZK_TIMEOUT,
+        Property.INSTANCE_SECRET, Property.INSTANCE_VOLUMES, Property.GENERAL_THREADPOOL_SIZE,
+        Property.GENERAL_DELEGATION_TOKEN_LIFETIME,
+        Property.GENERAL_DELEGATION_TOKEN_UPDATE_INTERVAL, Property.GENERAL_IDLE_PROCESS_INTERVAL,
+        Property.GENERAL_LOW_MEM_DETECTOR_INTERVAL, Property.GENERAL_LOW_MEM_DETECTOR_THRESHOLD,
+        Property.GENERAL_PROCESS_BIND_ADDRESS, Property.GENERAL_SERVER_LOCK_VERIFICATION_INTERVAL,
+        Property.MANAGER_CLIENTPORT, Property.TSERV_CLIENTPORT, Property.GC_CYCLE_START,
+        Property.GC_CYCLE_DELAY, Property.GC_PORT, Property.MONITOR_PORT, Property.TABLE_MAJC_RATIO,
+        Property.TABLE_SPLIT_THRESHOLD);
```
As the code is [currently written](https://github.com/apache/accumulo/blob/9e54d62244965d56994d4c74a6005aa0d7027411/core/src/main/java/org/apache/accumulo/core/conf/SiteConfiguration.java#L238-L248) it loops over all the props the loop in the prev comment may not work well because the get method replaces w/ the default value when not present.

In general it seems like it would be best to move the concept of a required property into the Property class in some form.  Then the entire system could react appropriately when a required property is not present and is requested.  For now a list in this class seems fine.

I experimented w/ validating the volume prop in #5365  based on the exploration done as part of this.

kevinrr888 replied:
This sounds like a good idea to me. There are a lot of `PropertyType.STRING` when a String isn't really what the property is. `PropertyType` exists to check that the property is valid; setting the `PropertyType` to `STRING` is just a way to ignore this validation. I wonder if it would be best for a 1 to 1 mapping `Property` to `PropertyType`. This would probably be overkill though, another idea could be to analyze the `PropertyType`s that are always valid. From briefly looking, `PropertyType.PATH`, `PropertyType.STRING`, `PropertyType.URI` are always valid. I don't think any properties should always be valid. Those that are `PropertyType.STRING` could probably be given a more appropriate existing `PropertyType` or a new one. `PATH` and `URI` could have validation.

In addition to this, can analyze all properties, determine if they are required or not, and change the validation:
- Properties that are not required could accept empty string, null, or a valid value (where validity is well defined)
- Properties that are required could only accept a valid value

Like you said, for this PR, can just push this list of required properties into `Property`. Maybe for now/in this PR this list of required props is only accessed/checked in this `admin check`. Might be a bit of a scope creep to start checking this list elsewhere in the code. Could do it in follow on.

keith-turner replied:
Definitely want to avoid any scope creep in this PR.  Identified some areas that need improvement based on this work, we can open follow on issues or PRs for those.

keith-turner replied:
Could leave the list as is in the PR. For follow on issues, do we need two issues?  One for addressing the STRING types and another for somehow representing and documenting required props in the Property.java?

kevinrr888 replied:
I think either would be fine, but it might be better as just one issue. There would be overlap in these changes so might be hard to split up/work on as two separate issues/PRs. For example, `instance.volumes` would need to be a required property and would also need to be moved away from `PropertyType.STRING`

kevinrr888 commented on server/base/src/main/java/org/apache/accumulo/server/util/checkCommand/SystemConfigCheckRunner.java:
```diff
@@ -18,22 +18,230 @@
 public class SystemConfigCheckRunner implements CheckRunner {
   private static final Admin.CheckCommand.Check check = Admin.CheckCommand.Check.SYSTEM_CONFIG;
 
+  public enum ServerProcess {
+    MANAGER, GC, TSERVER, COMPACTION_COORDINATOR, COMPACTOR, MONITOR, SCAN_SERVER
+  }
```
Is there anything in 3.x and 4.x?

kevinrr888 replied:
Since this now targets 4.x, no longer applicable. Using ServerId.Type now

meatballspaghetti commented on server/base/src/main/java/org/apache/accumulo/server/util/checkCommand/SystemConfigCheckRunner.java:
```diff
@@ -18,22 +18,205 @@
+  private static Admin.CheckCommand.CheckStatus checkZKLocks(ServerContext context,
+      Admin.CheckCommand.CheckStatus status) throws Exception {
+    final ServerProcess[] serverProcesses = ServerProcess.values();
+    final String zkRoot = context.getZooKeeperRoot();
```
Just noting now that if #5350 is merged before this PR, then the ZooKeeper root when using ServerContext will be moved to after `ZROOT+instanceId`. When that happens, the uses of `zkRoot` and `getZooKeeperRoot()` here will need to be omitted.

You can disregard this if this PR happens to be merged before #5350 though.

kevinrr888 replied:
Yeah, thanks for letting me know. I've been keeping that PR in mind in relation to this one.

kevinrr888 commented on server/base/src/main/java/org/apache/accumulo/server/util/checkCommand/RootTableCheckRunner.java:
```diff
@@ -51,8 +51,7 @@ public TableId tableId() {
-        MetadataSchema.TabletsSection.ServerColumnFamily.TIME_COLUMN,
-        MetadataSchema.TabletsSection.ServerColumnFamily.LOCK_COLUMN);
+        MetadataSchema.TabletsSection.ServerColumnFamily.TIME_COLUMN);
   }
```
Pushed into super

kevinrr888 commented:
@keith-turner - Could you take another look at this when you get the chance? I've updated this to work for 4.0. Since there's no more 3.x, this should make things simpler here. Main things changed in fbfa1fc89d2f681c39a68a1a107323e20d7e091d were:
* Addressed https://github.com/apache/accumulo/pull/5348#discussion_r1965737060
* `SystemConfigCheckRunner` changes

There are also some outstanding comments/questions

keith-turner commented on server/base/src/main/java/org/apache/accumulo/server/util/checkCommand/SystemConfigCheckRunner.java:
```diff
@@ -30,10 +46,189 @@ public Admin.CheckCommand.CheckStatus runCheck(ServerContext context, ServerUtil
+        // should be able to parse the WAL metadata
+        final var fullWalPath = tserverPath + "/" + walPath;
+        log.trace("Attempting to parse WAL metadata at {}", fullWalPath);
+        var parseRes = WalStateManager.parse(zrw.getData(fullWalPath));
```
There is a race condition where something could be deleted between calling getChildren() and getData().

```suggestion
        var data = zrw.getData(fullWalPath);
        if(data == null){
          continue;
        }
        var parseRes = WalStateManager.parse(data);
```
